src/dfx/src/lib/migrate.rs

```diff
@@ -0,0 +1,143 @@
+use ic_utils::{
+    interfaces::{
+        management_canister::builders::{CanisterSettings, InstallMode},
+        ManagementCanister, WalletCanister,
+    },
+    Argument,
+};
+use itertools::Itertools;
+
+use crate::{
+    lib::waiter::waiter_with_timeout,
+    util::{assets::wallet_wasm, expiry_duration},
+};
+
+use super::{
+    environment::Environment,
+    error::DfxResult,
+    identity::{Identity, IdentityManager},
+    models::canister_id_store::CanisterIdStore,
+    network::network_descriptor::NetworkDescriptor,
+    root_key::fetch_root_key_if_needed,
+};
+
+pub async fn migrate(env: &dyn Environment, network: &NetworkDescriptor, fix: bool) -> DfxResult {
+    fetch_root_key_if_needed(env).await?;
+    let config = env.get_config_or_anyhow()?;
+    let config = config.get_config();
+    let agent = env
+        .get_agent()
+        .expect("Could not get agent from environment");
+    let mut mgr = IdentityManager::new(env)?;
+    let ident = mgr.instantiate_selected_identity()?;
+    let mut did_migrate = false;
+    let wallet = Identity::wallet_canister_id(env, network, ident.name())
+        .map_err(|_| anyhow!("No wallet found; nothing to do"))?;
+    let wallet = if let Ok(wallet) = WalletCanister::create(agent, wallet).await {
+        wallet
+    } else {
+        let cbor = agent
+            .read_state_canister_info(wallet, "controllers", false)
+            .await?;
+        let controllers: Vec<Principal> = serde_cbor::from_slice(&cbor)?;
+        bail!("This identity isn't a controller of the wallet. You need to be one of these principals to upgrade the wallet: {}", controllers.into_iter().join(", "))
+    };
+    did_migrate |= migrate_wallet(env, agent, &wallet, fix).await?;
+    if let Some(canisters) = &config.canisters {
+        let store = CanisterIdStore::for_network(network)?;
+        for name in canisters.keys() {
+            if !config.is_remote_canister(name, &network.name)? {
+                if let Some(id) = store.find(name) {
+                    did_migrate |= migrate_canister(agent, &wallet, id, name, &ident, fix).await?;
+                }
+            }
+        }
+    }
+    if did_migrate {
+        println!("You can also make all of these changes at once with the `dfx fix` command");
+    } else {
+        println!("No problems found");
+    }
+    Ok(())
+}
+
+async fn migrate_wallet(
+    env: &dyn Environment,
+    agent: &Agent,
+    wallet: &WalletCanister<'_>,
+    fix: bool,
+) -> DfxResult<bool> {
+    let mgmt = ManagementCanister::create(agent);
+    if !wallet.version_supports_u128_cycles() {
+        if fix {
+            println!("Upgrading wallet... ");
+            let wasm = wallet_wasm(env.get_logger())?;
+            mgmt.install_code(wallet.canister_id_(), &wasm)
+                .with_mode(InstallMode::Upgrade)
+                .call_and_wait(waiter_with_timeout(expiry_duration()))
+                .await
+                .context("Could not upgrade wallet")?;
```

Comment:
There's a piece missing here: storing the new wallet wasm with the wallet itself so that the wallet's `wallet_create_wallet` call (which can be called through the wallet's ui) can create a wallet with the new wallet version.  This is taken from the 'dfx wallet upgrade' code: 
```
    wallet
        .wallet_store_wallet_wasm(wasm)
        .call_and_wait(waiter_with_timeout(expiry_duration()))
        .await?;
```

Reply:
For extra credit, refactor this and `dfx wallet upgrade` so that they share code.